src/v/raft/state_machine_manager.cc

```diff
@@ -185,10 +192,29 @@ ss::future<> state_machine_manager::start() {
     });
     std::vector<model::offset> offsets;
     for (const auto& [name, stm_meta] : _machines) {
+        vlog(
+          _log.trace,
+          "state machine {} last applied offset: {}",
+          name,
+          stm_meta->stm->last_applied_offset());
         offsets.push_back(stm_meta->stm->last_applied_offset());
     }
-    std::sort(offsets.begin(), offsets.end());
-    _next = model::next_offset(offsets.front());
+    _next = model::next_offset(*std::ranges::max_element(offsets));
+    auto log_offsets = _raft->log()->offsets();
+    /**
+     * Special case for the `archival_metadata_stm` local snapshot created
+     * during recovery. The snapshot last included offset was set to the log
+     * start offset.
```

Comment:
Thanks for the more detailed comment, it made me realise I don't understand this special case. The condition below holds iff the log is empty. Why is archival_metadata_stm local snapshot special so it is not consistent with the log position?

Reply:
yes, exactly